src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/MetricItemExtensions.cs

```diff
@@ -1,418 +1,434 @@
+
+namespace OpenTelemetry.Exporter.OpenTelemetryProtocol.Implementation;
+
+internal static class MetricItemExtensions
+{
+    private static readonly ConcurrentBag<OtlpMetrics.ScopeMetrics> MetricListPool = new();
+
+    internal static void AddMetrics(
+        this OtlpCollector.ExportMetricsServiceRequest request,
+        OtlpResource.Resource processResource,
+        in Batch<Metric> metrics)
+    {
+        var metricsByLibrary = new Dictionary<string, OtlpMetrics.ScopeMetrics>();
+        var resourceMetrics = new OtlpMetrics.ResourceMetrics
+        {
+            Resource = processResource,
+        };
+        request.ResourceMetrics.Add(resourceMetrics);
+
+        foreach (var metric in metrics)
+        {
+            var otlpMetric = metric.ToOtlpMetric();
+
+            // TODO: Replace null check with exception handling.
+            if (otlpMetric == null)
+            {
+                OpenTelemetryProtocolExporterEventSource.Log.CouldNotTranslateMetric(
+                    nameof(MetricItemExtensions),
+                    nameof(AddMetrics));
+                continue;
+            }
+
+            var meterName = metric.MeterName;
+            if (!metricsByLibrary.TryGetValue(meterName, out var scopeMetrics))
+            {
+                scopeMetrics = GetMetricListFromPool(meterName, metric.MeterVersion, metric.MeterTags);
+
+                metricsByLibrary.Add(meterName, scopeMetrics);
+                resourceMetrics.ScopeMetrics.Add(scopeMetrics);
+            }
+
+            scopeMetrics.Metrics.Add(otlpMetric);
+        }
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    internal static void Return(this OtlpCollector.ExportMetricsServiceRequest request)
+    {
+        var resourceMetrics = request.ResourceMetrics.FirstOrDefault();
+        if (resourceMetrics == null)
+        {
+            return;
+        }
+
+        foreach (var scope in resourceMetrics.ScopeMetrics)
+        {
+            scope.Metrics.Clear();
+            MetricListPool.Add(scope);
+        }
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    internal static OtlpMetrics.ScopeMetrics GetMetricListFromPool(string name, string version, IEnumerable<KeyValuePair<string, object>> meterTags)
+    {
+        if (!MetricListPool.TryTake(out var metrics))
+        {
+            metrics = new OtlpMetrics.ScopeMetrics
+            {
+                Scope = new OtlpCommon.InstrumentationScope
+                {
+                    Name = name, // Name is enforced to not be null, but it can be empty.
+                    Version = version ?? string.Empty, // NRE throw by proto
+                },
+            };
+            metrics.Scope.Attributes.Clear();
+            AddAttributes(meterTags, metrics.Scope.Attributes);
+        }
+        else
+        {
+            metrics.Scope.Name = name;
+            metrics.Scope.Version = version ?? string.Empty;
+            metrics.Scope.Attributes.Clear();

```

Comment:
This could be moved to line `82` to keep all clean up together?

Reply:
https://github.com/open-telemetry/opentelemetry-dotnet/issues/5090 Will address this separately to unblock the merge and release asap.